File: android/src/main/java/com/reactnativehypertracksdk/common/Serialization.kt
package com.reactnativehypertracksdk.common

import android.location.Location

/**
 * Platform-independent serialization code that converts HyperTrack data types
 * to Map<String, T> or List<T> where T is any JSON-compatible type
 */
internal object Serialization {

    fun serializeErrors(errors: Set<HyperTrackError>): List<Map<String, String>> {
        return errors.map {
            serializeHypertrackError(it)
        }
    }

    fun serializeLocationSuccess(location: Location): Map<String, Any?> {
        return serializeSuccess(serializeLocation(location))
    }

    fun serializeLocationWithDeviationSuccess(
        location: Location,
        deviation: Double
    ): Map<String, Any?> {
        return serializeSuccess(serializeLocationWithDeviation(
            location,
            deviation
        ))
    }

    fun serializeLocationErrorFailure(locationError: LocationError): Map<String, Any?> {
        return serializeFailure(serializeLocationError(locationError))
    }

    fun serializeIsTracking(isTracking: Boolean): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_IS_TRACKING,
            KEY_VALUE to isTracking
        )
    }

    fun serializeIsAvailable(isAvailable: Boolean): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_AVAILABILITY,
            KEY_VALUE to isAvailable
        )
    }

    fun serializeHypertrackError(error: HyperTrackError): Map<String, String> {
        return mapOf(
            KEY_TYPE to TYPE_HYPERTRACK_ERROR,
            KEY_VALUE to error.name
        )
    }

    fun serializeDeviceId(deviceId: String): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_DEVICE_ID,
            KEY_VALUE to deviceId
        )
    }

    fun deserializeDeviceName(name: Map<String, Any?>): Result<String> {
        return parse(name) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_DEVICE_NAME)
            it
                .get<String>(KEY_VALUE)
                .getOrThrow()
        }
    }

    fun deserializeAvailability(isAvailable: Map<String, Any?>): Result<Boolean> {
        return parse(isAvailable) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_AVAILABILITY)
            it
                .get<Boolean>(KEY_VALUE)
                .getOrThrow()
        }
    }

    fun deserializeGeotagData(map: Map<String, Any?>): Result<GeotagData> {
        return parse(map) {
            val data = it
                .get<Map<String, Any?>>(KEY_GEOTAG_DATA)
                .getOrThrow()
            val locationData = it
                .getOptional<Map<String, Any?>>(KEY_GEOTAG_EXPECTED_LOCATION)
                .getOrThrow()
            val location = locationData?.let { deserializeLocation(it).getOrThrow() }
            GeotagData(data, location)
        }
    }

    fun <T> parse(
        source: Map<String, Any?>,
        parseFunction: (Parser) -> T
    ): Result<T> {
        val parser = Parser(source)
        return try {
            if (parser.exceptions.isEmpty()) {
                Success(parseFunction.invoke(parser))
            } else {
                Failure(ParsingExceptions(source, parser.exceptions))
            }
        } catch (e: Exception) {
            Failure(
                if (parser.exceptions.isNotEmpty()) {
                    ParsingExceptions(source, parser.exceptions + e)
                } else {
                    e
                }
            )
        }
    }

    internal class Parser(
        private val source: Map<String, Any?>
    ) {
        private val _exceptions = mutableListOf<Exception>()
        val exceptions: List<Exception> = _exceptions

        inline fun <reified T> get(
            key: String
        ): Result<T> {
            return try {
                Success(source[key]!! as T)
            } catch (e: Exception) {
                Failure(
                    ParsingException(key, e)
                        .also {
                            _exceptions.add(it)
                        }
                )
            }
        }

        inline fun <reified T> getOptional(
            key: String
        ): Result<T?> {
            return try {
                Success(source[key] as T?)
            } catch (e: Exception) {
                Failure(
                    ParsingException(key, e)
                        .also {
                            _exceptions.add(it)
                        }
                )
            }
        }

        inline fun <reified T> assertValue(
            key: String,
            value: Any
        ) {
            if (source[key] != value) {
                _exceptions.add(Exception("Assertion failed: $key != $value"))
            }
        }
    }

    internal data class ParsingExceptions(
        val source: Any,
        val exceptions: List<Exception>
    ) : Throwable(
        exceptions.joinToString("\n")
            .let {
                "Invalid input:\n\n${source}\n\n$it"
            }
    )

    internal class ParsingException(
        key: String,
        exception: Exception
    ) : Exception("Invalid value for '$key': $exception", exception)

    private fun deserializeLocation(map: Map<String, Any?>): Result<Location> {
        return parse(map) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_LOCATION)
            val value = it
                .get<Map<String, Any?>>(KEY_VALUE)
                .getOrThrow()
            parse(value) { parser ->
                val latitude = parser
                    .get<Double>(KEY_LATITUDE)
                    .getOrThrow()
                val longitude = parser
                    .get<Double>(KEY_LONGITUDE)
                    .getOrThrow()
                Location("api").also {
                    it.latitude = latitude
                    it.longitude = longitude
                }
            }.getOrThrow()
        }
    }

    private fun serializeLocationWithDeviation(
        location: Location,
        deviation: Double
    ): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_LOCATION_WITH_DEVIATION,
            KEY_VALUE to mapOf(
                KEY_LOCATION to serializeLocation(location),
                KEY_DEVIATION to deviation
            )
        )
    }

    private fun serializeFailure(failure: Map<String, Any?>): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_RESULT_FAILURE,
            KEY_VALUE to failure
        )
    }

    private fun serializeSuccess(success: Map<String, Any?>): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_RESULT_SUCCESS,
            KEY_VALUE to success
        )
    }

    private fun serializeLocation(location: Location): Map<String, Any?> {
        return mapOf(
            KEY_TYPE to TYPE_LOCATION,
            KEY_VALUE to mapOf(
                KEY_LATITUDE to location.latitude,
                KEY_LONGITUDE to location.longitude
            )
        )
    }

    private fun serializeLocationError(locationError: LocationError): Map<String, Any?> {
        return when (locationError) {
            NotRunning -> {
                mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_NOT_RUNNING)
            }
            Starting -> {
                mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_STARTING)
            }
            is Errors -> {
                mapOf(
                    KEY_TYPE to TYPE_LOCATION_ERROR_ERRORS,
                    KEY_VALUE to locationError.errors
                        .map { serializeHypertrackError(it) }
                )
            }
        }
    }

    private const val KEY_TYPE = "type"
    private const val KEY_VALUE = "value"

    private const val TYPE_RESULT_SUCCESS = "success"
    private const val TYPE_RESULT_FAILURE = "failure"

    private const val TYPE_LOCATION = "location"
    private const val TYPE_LOCATION_WITH_DEVIATION = "locationWithDeviation"
    private const val TYPE_AVAILABILITY = "isAvailable"
    private const val TYPE_DEVICE_NAME = "deviceName"
    private const val TYPE_DEVICE_ID = "deviceID"
    private const val TYPE_HYPERTRACK_ERROR = "hyperTrackError"
    private const val TYPE_IS_TRACKING = "isTracking"

    private const val TYPE_LOCATION_ERROR_NOT_RUNNING = "notRunning"
    private const val TYPE_LOCATION_ERROR_STARTING = "starting"
    private const val TYPE_LOCATION_ERROR_ERRORS = "errors"

    private const val KEY_LATITUDE = "latitude"
    private const val KEY_LONGITUDE = "longitude"

    const val KEY_GEOTAG_DATA = "data"
    const val KEY_GEOTAG_EXPECTED_LOCATION = "expectedLocation"
    const val KEY_DEVIATION = "deviation"
    const val KEY_LOCATION = "location"
}

File: ios/common/Serialization.swift
import HyperTrack

private let keyType = "type"
private let keyValue = "value"
private let keyGeotagData = "data"
private let keyExpectedLocation = "expectedLocation"

private let typeSuccess = "success"
private let typeFailure = "failure"
private let typeHyperTrackError = "hyperTrackError"
private let typeIsTracking = "isTracking"
private let typeIsAvailable = "isAvailable"
private let typeLocation = "location"

func getJSON(_ dictionary: [String: Any]) -> HyperTrack.JSON.Object {
    return try! DictionaryDecoder().decode(HyperTrack.JSON.Object.self, from: dictionary)
}

func getHyperTrackError(_ error: HyperTrack.UnrestorableError) -> HyperTrackError {
    switch error {
    case .invalidPublishableKey:
        return HyperTrackError.invalidPublishableKey
    case .motionActivityPermissionsDenied:
        return HyperTrackError.locationPermissionsDenied
    }
}

func getHyperTrackError(_ error: HyperTrack.RestorableError) -> HyperTrackError {
    switch error {
    case .locationPermissionsNotDetermined:
        return HyperTrackError.locationPermissionsNotDetermined
    case .motionActivityPermissionsNotDetermined:
        return HyperTrackError.motionActivityPermissionsNotDetermined
    case .locationPermissionsCantBeAskedInBackground:
        return HyperTrackError.locationPermissionsNotDetermined
    case .motionActivityPermissionsCantBeAskedInBackground:
        return HyperTrackError.motionActivityPermissionsNotDetermined
    case .locationPermissionsRestricted:
        return HyperTrackError.locationPermissionsRestricted
    case .motionActivityPermissionsRestricted:
        return HyperTrackError.motionActivityPermissionsRestricted
    case .locationPermissionsDenied:
        return HyperTrackError.locationPermissionsDenied
    case .locationPermissionsInsufficientForBackground:
        return HyperTrackError.locationPermissionsInsufficientForBackground
    case .locationServicesDisabled:
        return HyperTrackError.locationServicesDisabled
    case .motionActivityServicesDisabled:
        return HyperTrackError.motionActivityServicesDisabled
    case .networkConnectionUnavailable:
        return HyperTrackError.networkConnectionUnavailable
    case .trialEnded:
        return HyperTrackError.blockedFromRunning
    case .paymentDefault:
        return HyperTrackError.blockedFromRunning
    }
}

func getHyperTrackError(_ error: HyperTrack.HyperTrackError) -> HyperTrackError {
    switch error {
    case .locationPermissionsNotDetermined:
        return .locationPermissionsNotDetermined
    case .locationPermissionsInsufficientForBackground:
        return .locationPermissionsInsufficientForBackground
    case .locationPermissionsRestricted:
        return .locationPermissionsRestricted
    case .locationPermissionsReducedAccuracy:
        return .locationPermissionsReducedAccuracy
    case .locationPermissionsProvisional:
        return .locationPermissionsProvisional
    case .locationPermissionsDenied:
        return .locationPermissionsDenied
    case .locationServicesDisabled:
        return .locationServicesDisabled
    case .motionActivityPermissionsNotDetermined:
        return .motionActivityPermissionsNotDetermined
    case .motionActivityPermissionsDenied:
        return .motionActivityPermissionsDenied
    case .motionActivityServicesDisabled:
        return .motionActivityServicesDisabled
    case .motionActivityServicesUnavailable:
        return .motionActivityServicesUnavailable
    case .gpsSignalLost:
        return .gpsSignalLost
    case .locationMocked:
        return .locationMocked
    case .invalidPublishableKey:
        return .invalidPublishableKey
    case .blockedFromRunning:
        return .blockedFromRunning
    default:
        preconditionFailure("Unknown error \(error)")
    }
}

func serializeErrors(_ errors: Set<HyperTrack.HyperTrackError>) -> [[String: Any]] {
    return errors.map { error in
        serializeHyperTrackError(getHyperTrackError(error))
    }
}

func serializeDeviceID(_ deviceID: String) -> [String: Any] {
    return [
        keyType: "deviceID",
        keyValue: deviceID,
    ]
}

func serializeLocationResult(_ result: Result<HyperTrack.Location, HyperTrack.LocationError>) -> [String: Any] {
    switch result {
    case let .success(success):
        return [
            keyType: typeSuccess,
            keyValue: [
                keyType: typeLocation,
                keyValue: [
                    "latitude": success.latitude,
                    "longitude": success.longitude,
                ],
            ],
        ]
    case let .failure(failure):
        var locationError: [String: Any]
        switch failure {
        case .locationPermissionsNotDetermined:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsNotDetermined)
        case .locationPermissionsCantBeAskedInBackground:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsNotDetermined)
        case .locationPermissionsInsufficientForBackground:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsInsufficientForBackground)
        case .locationPermissionsRestricted:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsRestricted)
        case .locationPermissionsReducedAccuracy:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsReducedAccuracy)
        case .locationPermissionsDenied:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsDenied)
        case .locationServicesDisabled:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationServicesDisabled)
        case .motionActivityPermissionsNotDetermined:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityPermissionsNotDetermined)
        case .motionActivityPermissionsCantBeAskedInBackground:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityPermissionsNotDetermined)
        case .motionActivityPermissionsDenied:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityPermissionsDenied)
        case .motionActivityServicesDisabled:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityServicesDisabled)
        case .gpsSignalLost:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.gpsSignalLost)
        case .locationMocked:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationMocked)
        case .starting:
            locationError = [
                keyType: "starting",
            ]
        case .notRunning:
            locationError = [
                keyType: "notRunning",
            ]
        }
        return [
            keyType: typeFailure,
            keyValue: locationError,
        ]
    }
}

private func serializeLocation(_ location: HyperTrack.Location) -> [String: Any] {
    return [
        keyType: typeLocation,
        keyValue: [
            "latitude": location.latitude,
            "longitude": location.longitude,
        ],
    ]
}

func serializeLocationResult(_ result: Result<HyperTrack.Location, HyperTrack.NoLocationError>) -> [String: Any] {
    switch result {
    case let .success(success):
        return [
            keyType: typeSuccess,
            keyValue: serializeLocation(success),
        ]
    case let .failure(failure):
        var locationError: [String: Any]
        switch failure {
        case .starting:
            locationError = [
                keyType: "starting",
            ]
        case .notRunning:
            locationError = [
                keyType: "notRunning",
            ]
        case let .errors(errors):
            locationError = [
                keyType: "errors",
                keyValue: serializeErrors(errors),
            ]
        }

        return [
            keyType: typeFailure,
            keyValue: locationError,
        ]
    }
}

func serializeExpectedLocationResult(
    _ result: Result<HyperTrack.LocationWithDeviation, HyperTrack.NoLocationError>
) -> [String: Any] {
    switch result {
    case let .success(success):
        return [
            keyType: typeSuccess,
            keyValue: [
                keyType: "locationWithDeviation",
                keyValue: serializeLocationWithDeviation(success),
            ],
        ]
    case let .failure(failure):
        return [
            keyType: typeFailure,
            keyValue: serializeLocationError(failure),
        ]
    }
}

func serializeLocationError(_ error: HyperTrack.NoLocationError) -> [String: Any] {
    switch error {
    case .starting:
        return [
            keyType: "starting",
        ]
    case .notRunning:
        return [
            keyType: "notRunning",
        ]
    case let .errors(errors):
        return [
            keyType: "errors",
            keyValue: serializeErrors(errors),
        ]
    }
}

func serializeLocationWithDeviation(
    _ locationWithDeviation: HyperTrack.LocationWithDeviation
) -> [String: Any] {
    return [
        "location": serializeLocation(locationWithDeviation.location),
        "deviation": locationWithDeviation.deviation,
    ]
}

func serializeHyperTrackErrorAsLocationError(_ error: HyperTrackError) -> [String: Any] {
    return [
        keyType: "errors",
        keyValue: [
            [
                keyType: typeHyperTrackError,
                keyValue: error.rawValue,
            ],
        ],
    ]
}

func serializeHyperTrackError(_ error: HyperTrackError) -> [String: Any] {
    return [
        keyType: typeHyperTrackError,
        keyValue: error.rawValue,
    ]
}

func serializeIsTracking(_ isTracking: Bool) -> [String: Any] {
    return [
        keyType: typeIsTracking,
        keyValue: isTracking,
    ]
}

func serializeIsAvailable(_ isAvailable: HyperTrack.Availability) -> [String: Any] {
    return [
        keyType: typeIsAvailable,
        keyValue: isAvailable == .available,
    ]
}

func deserializeDeviceName(_ data: [String: Any]) -> Result<String, FailureResult> {
    if data[keyType] as? String != "deviceName" {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? String else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeAvailability(_ data: [String: Any]) -> Result<Bool, FailureResult> {
    if data[keyType] as? String != typeIsAvailable {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? Bool else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeGeotagData(
    _ args: [String: Any]
) -> Result<GeotagData, FailureResult> {
    guard let data = args[keyGeotagData] as? [String: Any] else {
        return .failure(.fatalError(getParseError(args, key: keyGeotagData)))
    }
    if let expectedLocationData = args[keyExpectedLocation] as? [String: Any] {
        let expectedLocation = deserializeLocation(expectedLocationData)
        switch expectedLocation {
        case let .failure(failure):
            return .failure(failure)
        case let .success(expectedLocation):
            return .success(.init(data: data, expectedLocation: expectedLocation))
        }
    } else {
        return .success(.init(data: data, expectedLocation: nil))
    }
}

func deserializeLocation(_ data: [String: Any]) -> Result<HyperTrack.Location, FailureResult> {
    if data[keyType] as? String != typeLocation {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? [String: Any] else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    guard let latitude = value["latitude"] as? Double else {
        return .failure(.fatalError(getParseError(value, key: "latitude")))
    }
    guard let longitude = value["longitude"] as? Double else {
        return .failure(.fatalError(getParseError(value, key: "longitude")))
    }

    return .success(HyperTrack.Location(latitude: latitude, longitude: longitude))
}

func getParseError(_ data: Any, key: String) -> String {
    return "Invalid input for key \(key): \(data)"
}
